hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngrade.java

```diff
@@ -0,0 +1,653 @@
+    
+    // Validate properties for target version
+    switch (toVersion) {
+      case FOUR:
+        validateVersion4Properties(tableConfig);
+        break;
+      case FIVE:
+        validateVersion5Properties(tableConfig);
+        break;
+      case SIX:
+        validateVersion6Properties(metaClient);
+        break;
+      case EIGHT:
+        validateVersion8Properties(tableConfig);
+        break;
+      case NINE:
+        validateVersion9Properties(metaClient, tableConfig);
+        break;
+      default:
+        LOG.warn("No specific property validation for version {}", toVersion);
+    }
+    
+  }
+
+  /**
+   * Validate specific properties for a given table version.
+   */
+  private void validateSpecificPropertiesForVersion(HoodieTableMetaClient metaClient, HoodieTableConfig tableConfig, HoodieTableVersion version) throws IOException {
+    switch (version) {
+      case FOUR:
+        validateVersion4Properties(tableConfig);
+        break;
+      case FIVE:
+        validateVersion5Properties(tableConfig);
+        break;
+      case SIX:
+        validateVersion6Properties(metaClient);
+        break;
+      case EIGHT:
+        validateVersion8Properties(tableConfig);
+        break;
+      case NINE:
+        validateVersion9Properties(metaClient, tableConfig);
+        break;
+      default:
+        LOG.warn("No specific property validation for version {}", version);
+    }
+  }
+
+  /**
+   * Validate basic properties for version 4.
+   */
+  private void validateVersion4Properties(HoodieTableConfig tableConfig) {
+    // TABLE_CHECKSUM should exist and be valid
+    assertTrue(tableConfig.contains(HoodieTableConfig.TABLE_CHECKSUM),
+        "TABLE_CHECKSUM should be set for V4");
+    String actualChecksum = tableConfig.getString(HoodieTableConfig.TABLE_CHECKSUM);
+    assertNotNull(actualChecksum, "TABLE_CHECKSUM should not be null");
+    
+    // Validate that the checksum is valid by comparing with computed checksum
+    String expectedChecksum = String.valueOf(HoodieTableConfig.generateChecksum(tableConfig.getProps()));
+    assertEquals(expectedChecksum, actualChecksum, 
+        "TABLE_CHECKSUM should match computed checksum");
+    
+    // TABLE_METADATA_PARTITIONS should be properly set if present
+    // Note: This is optional based on whether metadata table was enabled during upgrade
+    if (tableConfig.contains(HoodieTableConfig.TABLE_METADATA_PARTITIONS)) {
+      String metadataPartitions = tableConfig.getString(HoodieTableConfig.TABLE_METADATA_PARTITIONS);
+      assertTrue(metadataPartitions.contains("files"), 
+          "TABLE_METADATA_PARTITIONS should contain 'files' partition when set");
+    }
+  }
+
+  /**
+   * Validate properties for version 5 (default partition path migration).
+   */
+  private void validateVersion5Properties(HoodieTableConfig tableConfig) {
+    // Version 5 upgrade validates that no deprecated default partition paths exist
```

Comment:
Still call `validateVersion4Properties` as part of `validateVersion5Properties` like before?  Similarly for other version-specific validation if that applies.

Reply:
For now im going to assume then that `validateVersion6Properties` should call validateVersion5Properties, and `validateVersion5Properties` will call `validateVersion4Properties`

Simarily for `validateVersion9Properties` it will now call `validateVersion8Properties`.